Add direction classification algorithm for all connections

stb2132 opened the pull request:
### What does this PR do?

This PR adds an algorithm to determine connection direction by utilizing the current state of all connections. Since we use direction to determine if we should skip a connection, it also refactors code for skipping connection out of the getActiveConnections() and getInitPerfPolling() functions into the getConnections() function. 

### Motivation

We currently cannot properly classify connections as local since we do not compare a connection to the others, so these changes allow us to do this classification. 

### Additional Notes

This change will change some DNS connections from being classified as `NONE` to `LOCAL`. For  example, previously a connection with addresses like this:

```
 {
      "pid": 25991,
      "laddr": {
        "ip": "127.0.0.1",
        "port": 47995,
        "containerId": "",
        "hostId": 0
      },
      "raddr": {
        "ip": "127.0.0.1",
        "port": 8125,
        "containerId": "",
        "hostId": 0
      },
      "type": "udp",
      ...
    }
  ],
  "dns": {}
```

would have direction `NONE`, but will now be specified as `LOCAL`.

This change fixes the unit tests:

- TestTCPSendAndReceive
- TestPreexistingConnectionDirection 
- TestTCPOverIPv6
- TestLocalDNSCollectionDisabled 
- TestShouldSkipBlacklistedConnection

Commit:
Add direction classification algorithm

Algorithm works by:

- Accepting a slice of connections
- Creating a map of raddr:[]*connections
- Iterate through the laddr of connection slice
- If there is a hit in the map, classify the connection as Local
- Otherwise, classify as None/Incoming/Outgoing depending on connection
properties

Commit:
Refactor tracer logic to utilize new direction classification func

Commit:
Fix TestUDPSendAndReceive unit test to reflect properly classified conn direction

leeavital commented on pkg/ebpf/tracer.go:
```diff
@@ -644,16 +637,51 @@ func (t *Tracer) populatePortMapping(mp *bpflib.Map) ([]uint16, error) {
+}
 
-	return OUTGOING
+func (t *Tracer) removeExpiredConns(allConns []ConnectionStats) []ConnectionStats {
```
let's call this `removeBlacklistedConnections` since the connections being removed here are not expired.

leeavital commented on pkg/ebpf/tracer.go:
```diff
@@ -644,16 +637,51 @@ func (t *Tracer) populatePortMapping(mp *bpflib.Map) ([]uint16, error) {
+		if conn.IPTranslation == nil{
+			connRAddrMap[conn.Dest] = append(connRAddrMap[conn.Dest], &conn)
+		} else {
+			connRAddrMap[conn.IPTranslation.ReplDstIP] = append(connRAddrMap[conn.IPTranslation.ReplDstIP], &conn)
```
This should be `conn.IPTranslation.ReplSrcIP` because the fields for IPTranslation are in reverse

kevinconaway replied:
Should this be the `ReplSrcIP`? (cc @leeavital)

leeavital commented on pkg/ebpf/tracer.go:
```diff
@@ -644,16 +637,51 @@ func (t *Tracer) populatePortMapping(mp *bpflib.Map) ([]uint16, error) {
+			connections[i].Direction = LOCAL
+			for j := 0; j < len (hit); j++ {
+				hit[j].Direction = LOCAL
+			}
```
I don't think this part is necessary, everything in `hit` will end up getting processed once as part of the outer loop.   That means that the `connRAddrMap` can be a map from `util.Address` to `struct{}` which will save on space.

kevinconaway commented on pkg/ebpf/tracer.go:
```diff
@@ -644,16 +637,51 @@ func (t *Tracer) populatePortMapping(mp *bpflib.Map) ([]uint16, error) {
+		if connections[i].Direction == LOCAL{
+			continue
+		}
+		hit, ok := connRAddrMap[connections[i].Source]
```
nit, if this is a collection, would you please name it `hits` or perhaps something like `localRaddrs`?

stb2132 replied:
Thanks for the note - going to be removing the inner loop, but will keep this in mind moving forward.

leeavital commented on pkg/ebpf/tracer.go:
```diff
@@ -644,16 +637,51 @@ func (t *Tracer) populatePortMapping(mp *bpflib.Map) ([]uint16, error) {
+		if connections[i].Direction == LOCAL{
+			continue
+		}
+		hit, ok := connRAddrMap[connections[i].Source]
```
This part should also be IPtranslation aware, and use IPTranslation.ReplDstIP if available

kevinconaway commented on pkg/ebpf/tracer.go:
```diff
@@ -644,16 +637,51 @@ func (t *Tracer) populatePortMapping(mp *bpflib.Map) ([]uint16, error) {
-	if t.portMapping.IsListening(conn.SPort) {
-		return INCOMING
+	for i := 0; i < len(connections); i++ {
+		if connections[i].Direction == LOCAL{
```
nit, it might simplify things to alias `connections[i]` to a local var in this loop to prevent mistakes since you have an inner loop

kevinconaway reviewed (commented):
It doesn't look like `go fmt` was run on these files.  Would you please do so?

I'm surprised the agent builds without it.

Commit:
Update algorithm in response to feedback and fix formatting

sunhay commented on pkg/ebpf/tracer.go:
```diff
@@ -273,13 +274,9 @@ func (t *Tracer) initPerfPolling() (*bpflib.PerfMap, error) {
-					t.state.StoreClosedConnection(cs)
-				}
+				cs.IPTranslation = t.conntracker.GetTranslationForConn(cs.Source, cs.SPort, process.ConnectionType(cs.Type))
+				// Store all connections, we will determine if we should skip/if it is closed in GetConnections
```
Why? Do we know how many more connections we'll be storing in state? I've seen 10k+ local dns connections before, which I'd prefer to leave out of the state...

sunhay replied:
I wasn't aware that we were using direction to skip connections, when was that added?

leeavital replied:
This was moved because we now need the full list of connections to determine connection direction and we only have one available here.

shouldSkipConnection is defined as

```
func (t *Tracer) shouldSkipConnection(conn *ConnectionStats) bool {
	isDNSConnection := conn.DPort == 53 || conn.SPort == 53
	if !t.config.CollectLocalDNS && isDNSConnection && conn.Direction == LOCAL {
		return true
	} else if IsBlacklistedConnection(t.sourceExcludes, t.destExcludes, conn) {
		return true
	}
	return false
}
```

The only place it uses Local is to enfore the older CollectLocalDNS option which was replaced by connection blacklisting.  How about removing the `conn.Direction == LOCAL` check and going back to calling `shouldSkipConnection` before a connection gets stored in `t.state`

sunhay commented on pkg/ebpf/tracer.go:
```diff
@@ -644,16 +637,78 @@ func (t *Tracer) populatePortMapping(mp *bpflib.Map) ([]uint16, error) {
 
-	if t.portMapping.IsListening(conn.SPort) {
-		return INCOMING
+	newConnKey := func(connStat *ConnectionStats, RAddr bool) connKey {
```
`RAddr` -> `useRAddrAsKey` so you don't need the comment below

sunhay commented on pkg/ebpf/tracer.go:
```diff
@@ -644,16 +637,78 @@ func (t *Tracer) populatePortMapping(mp *bpflib.Map) ([]uint16, error) {
-	if t.portMapping.IsListening(conn.SPort) {
-		return INCOMING
+	newConnKey := func(connStat *ConnectionStats, RAddr bool) connKey {
+		key := connKey{}
```
`key := connKey{ Type: connStat.Type}`

sunhay commented on pkg/ebpf/tracer.go:
```diff
@@ -644,16 +637,78 @@ func (t *Tracer) populatePortMapping(mp *bpflib.Map) ([]uint16, error) {
+		keyWithRAddr = newConnKey(connToModify, true)
+		_, ok := connLAddrMap[keyWithRAddr]
+		if ok {
+			connToModify.Direction = LOCAL
```
I thought we were going to get rid of `Direction = LOCAL` as a concept?

sunhay commented on pkg/ebpf/tracer.go:
```diff
@@ -644,16 +637,78 @@ func (t *Tracer) populatePortMapping(mp *bpflib.Map) ([]uint16, error) {
+				connToModify.Direction = NONE
+				continue
+			}
+			if t.portMapping.IsListening(connToModify.SPort) {
```
nit: this should just be an `else if` on the above block

sunhay commented on pkg/ebpf/tracer.go:
```diff
@@ -644,16 +637,78 @@ func (t *Tracer) populatePortMapping(mp *bpflib.Map) ([]uint16, error) {
+	}
+
+	keyWithRAddr := connKey{}
+	for i := 0; i < len(connections); i++ {
```
`for _, i := range connections`?

sunhay commented on pkg/ebpf/tracer.go:
```diff
@@ -644,16 +637,78 @@ func (t *Tracer) populatePortMapping(mp *bpflib.Map) ([]uint16, error) {
+	connLAddrMap := make(map[connKey]struct{}, 0)
+	keyWithLAddr := connKey{}
+	// Create a map of RAddrs to their respective connections
+	for i := 0; i < len(connections); i++ {
```
`for _, i := range connections`

leeavital replied:
Or even:

```
for _, conn := range connections {
   connLadddrMap[newConnKey(conn, false)] = struct{}{}
}
```

Commit:
Add first round of unit tests

leeavital commented on pkg/ebpf/tracer.go:
```diff
@@ -644,16 +637,83 @@ func (t *Tracer) populatePortMapping(mp *bpflib.Map) ([]uint16, error) {
-	if t.portMapping.IsListening(conn.SPort) {
-		return INCOMING
+	connLAddrMap := make(map[connKey]struct{}, 0)
+	keyWithLAddr := connKey{}
```
I don't think this variable needs to be declared outside the for loop immediately below this line.

leeavital commented on pkg/ebpf/tracer.go:
```diff
@@ -644,16 +637,83 @@ func (t *Tracer) populatePortMapping(mp *bpflib.Map) ([]uint16, error) {
 
-	if t.portMapping.IsListening(conn.SPort) {
-		return INCOMING
+	connLAddrMap := make(map[connKey]struct{}, 0)
```
we can call this `connLAddrs`  or even `laddrs`. No need to put `map` in the name since that's part of the type

leeavital commented on pkg/ebpf/tracer_test.go:
```diff
@@ -7,6 +7,7 @@ import (
 	"bytes"
 	"encoding/json"
 	"fmt"
+	"github.com/DataDog/datadog-agent/pkg/ebpf/netlink"
```
nit: put this import with the rest of the "github.com/DataDog/..." imporrts

leeavital commented on pkg/ebpf/tracer_test.go:
```diff
@@ -238,6 +240,91 @@ func TestPreexistingConnectionDirection(t *testing.T) {
+	connStats.SPort = 5323
+	connStats.Dest = util.AddressFromString("38.122.226.210")
+	connStats.DPort = 80
+	connStats.Type = TCP
```
nit: these are more clearly written as a literal

```
connStats := ConnectionStats{
   Source: util.AddressFromString("10.0.2.25"),
   Sport: 5223,
   Dest: ...
}
```

leeavital commented on pkg/ebpf/tracer_test.go:
```diff
@@ -238,6 +240,91 @@ func TestPreexistingConnectionDirection(t *testing.T) {
+
+func TestOutgoingTCPConnectionDirection(t *testing.T) {
+	tr, err := NewTracer(NewDefaultConfig())
+	if err != nil {
```
nit: this is better written as `require.NoErr(t, err)

leeavital commented on pkg/ebpf/tracer_test.go:
```diff
@@ -238,6 +240,91 @@ func TestPreexistingConnectionDirection(t *testing.T) {
+}
+
+func TestOutgoingTCPConnectionDirection(t *testing.T) {
+	tr, err := NewTracer(NewDefaultConfig())
```
nit: instead of using `NewTracer()`  which will attach a BPF program, you can do something like:

```go
tr := &tracer{
   // initialize anything necessary here
}
```

And move this to a test file that isn't tagged as `linux_bpf`. That way these tests will run as part of a normal `go test`

stb2132 replied:
Moved other tests to a file and do note enable a full tracer. This test remains as an integration test.

leeavital commented on pkg/ebpf/tracer.go:
```diff
@@ -644,16 +637,83 @@ func (t *Tracer) populatePortMapping(mp *bpflib.Map) ([]uint16, error) {
+		if ok {
+			undirectedConn.Direction = LOCAL
+		} else {
+			if undirectedConn.Type == UDP {
```
nit: I suggest writing this as an `if/else if/else` and avoiding `continue`

leeavital commented on pkg/ebpf/tracer_test.go:
```diff
@@ -238,6 +240,91 @@ func TestPreexistingConnectionDirection(t *testing.T) {
 	doneChan <- struct{}{}
 }
 
+func TestIncomingTCPConnectionDirection(t *testing.T) {
```
For these four similar tests, you may be able to use a table test to avoid copying the same setup across several tests.

Commit:
Update code based upon comments.

Main changes include:

- Add back shouldSkipConnection functionality on a per-conn basis
- Rename function variables in setConnectionDirections to make clearer

Commit:
Complete adding back shouldSkipConnection

Commit:
Add direction tests to separate file - Include potential table test object

Commit:
Remove potential table test object

Commit:
Add integration test outline

Commit:
Change error check from if statement to assertion

sunhay commented on pkg/ebpf/tracer.go:
```diff
@@ -304,7 +304,7 @@ func (t *Tracer) initPerfPolling() (*bpflib.PerfMap, error) {
 func (t *Tracer) shouldSkipConnection(conn *ConnectionStats) bool {
 	isDNSConnection := conn.DPort == 53 || conn.SPort == 53
-	if !t.config.CollectLocalDNS && isDNSConnection && conn.Direction == LOCAL {
+	if !t.config.CollectLocalDNS && isDNSConnection {
```
Hmm we'll want to make sure we add an additional check to confirm this is a local DNS connection if we're removing `conn.Direction == LOCAL`

sunhay replied:
Possibly looking at the IP to see if it's in the LAN subnets

leeavital replied:
https://github.com/DataDog/datadog-agent/pull/4727/files will fix this.

Commit:
Add protobuf field to agent repository

Commit:
Add field to have conntracker initialized properly in integration tests

Commit:
Fix stash changes merge conflict

Commit:
Add print statements for testing

Commit:
Add ability to pass in argument to invoke test to run a specific test

leeavital committed:
Fix tests

leeavital committed:
more stuff

Commit:
Finish algoirthm and testing changes

Commit:
Revert some changes

leeavital commented on pkg/ebpf/tracer.go:
```diff
@@ -656,6 +658,52 @@ func (t *Tracer) determineConnectionDirection(conn *ConnectionStats) ConnectionD
+		return key
+	}
+
+	lAddrs := make(map[connKey]struct{}, 0)
```
You can leave out the 0 here.

stb2132 replied:
Removed this - looks like it doesn't update here though. Resolving conversation.

leeavital reviewed (commented):
LGTM pending us referencing a tag of agent-payload, and not a hash

Commit:
Remove unnecessary 0 for declaration

kevinconaway commented on pkg/ebpf/tracer.go:
```diff
@@ -336,6 +337,7 @@ func (t *Tracer) GetActiveConnections(clientID string) (*Connections, error) {
 	}
 
 	conns := t.state.Connections(clientID, latestTime, latestConns)
+	t.determineConnectionIntraHost(conns)
```
nit, does this need to be done in the tracer or can it be done in `Connections` ?

leeavital replied:
Does this actually fit better in `state.Connections` ? `tracer` is already doing a bunch of post-processing (like  handling connection direction, ip translations, and DNS). This seems to fit in nicely.

kevinconaway replied:
🤷‍♂ my thought was that since the method does nothing specific with the tracer data (the method itself is stateless) it would be better off elsewhere so that other calls to `Connections` would have the data.  

I don't have strong feelings though.

stb2132 replied:
I've been playing around with this, and it can be done in `Connections` fairly simply. I have the changes on my local branch and will push them up.

kevinconaway commented on pkg/ebpf/tracer.go:
```diff
@@ -656,6 +658,52 @@ func (t *Tracer) determineConnectionDirection(conn *ConnectionStats) ConnectionD
+				key.Port = connStat.IPTranslation.ReplSrcPort
+			}
+		} else {
+			key.Address = connStat.Source
```
We don't need to take the IPTranslation in to account for the source IP?

leeavital replied:
No, because of how NAT connections show up for intra-host connections, the original IP is the "real" IP and the NATed source is the "virtual" IP.

Commit:
Fix formatting

Commit:
Move intrahost determination algorithm to state.go and update tests

kevinconaway reviewed (dismissed):
Before merging, please make a release of the agent-payload and update the PR with the actual version.

Commit:
Update Gopkg.toml to use a version instead of a revision

Commit:
Require test to be run against BPF tracer

Commit:
Update gopkg.lock file to correspond with gopkg.toml change

Commit:
Add error check to test

stb2132 commented:
Merged change in agent-payload:

- https://github.com/DataDog/agent-payload/releases/tag/4.24.0

derekwbrown reviewed (approved):
should I ask what happened to 20-23?
LGTM
